### PR TITLE
feature/devcontainer

### PR DESCRIPTION
**Pull request by marcellanz**
Add devcontainer support.

- based on a Scala/SBT image and enabled docker support.
- Tested on Github Codespaces with the TCK running.

We might provide a pre-built docker image so devcontainers get started faster.

Screenshot 2020-08-03 at 10 48 06

### CODE REVIEW

**Comment by viktorklang** [comment]
Thanks for contributing this, @marcellanz :+1:

**Comment by marcellanz** [comment]
The image can be built with:

```
> cd .devcontainer
> ./build.sh [push]
```
and add:
`"image": "cloudstateio/cloudstate-devcontainer:latest",`

to `devcontainer.json` instead of `"dockerFile": "Dockerfile",`

File: .devcontainer/devcontainer.json [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+  "dockerFile": "Dockerfile",
+  "settings": {
+    "terminal.integrated.shell.linux": "/bin/bash",
+    "[scala]": {
+      "editor.tabSize": 2,
+      "editor.insertSpaces": true
+    }
+  },
+  "extensions": [
+    "scala-lang.scala", "zxh404.vscode-proto3"
```
Comment (marcellanz) [inline .devcontainer/devcontainer.json:12]: I did not append more extensions here. One might install her preferred extensions (metals?) for Scala and the like through [`.dotfiles`](https://docs.microsoft.com/en-us/visualstudio/codespaces/reference/personalizing).

**Comment by pvlugter** [comment]
👍 Good idea.

Having pre-built docker image sounds good too — can be add to the travis build, which has docker hub credentials already. I've created the repo on docker hub for this, with permissions for you and bots.

File: .devcontainer/build.sh [hunk 2]
```diff
@@ -0,0 +1,11 @@
+#!/usr/bin/env bash
+
+set -o nounset
+set -o errexit
+set -o pipefail
+
+CMD=${1:-image}
+DOCKER_BUILDKIT=1 docker build . -t cloudstateio/cloudstate-devcontainer:latest -f Dockerfile
+if [ "$CMD" == "push" ]; then
+  docker push cloudstateios/cloudstate-devcontainer:latest
```
[the remark this answers is not in the document]
Comment (marcellanz) [reply .devcontainer/build.sh:10]: :) thanks. that was my guard to test my script but not to push.

**Comment by marcellanz** [comment]
@pvlugter thanks. I've published [`cloudstateio/cloudstate-devcontainer`](https://hub.docker.com/repository/docker/cloudstateio/cloudstate-devcontainer) and tested it. I Added the build and push script on "Publish latest builds" in the travis configuration. We probably might not have releases except of latest of this image?

**Comment by pvlugter** [comment]
> I've published [`cloudstateio/cloudstate-devcontainer`](https://hub.docker.com/repository/docker/cloudstateio/cloudstate-devcontainer) and tested it. I Added the build and push script on "Publish latest builds" in the travis configuration. We probably might not have releases except of latest of this image?

Cool. Yes, makes sense to just have latest builds for this.

**Comment by ralphlaude** [comment]
I test it locally with the TCK and it works fine

**Comment by marcellanz** [comment]
> I test it locally with the TCK and it works fine

thanks for testing @ralphlaude 👍

**Pull request merged**